libs/matdbg/src/ShaderReplacer.cpp

```diff
@@ -159,6 +200,93 @@ bool ShaderReplacer::replaceShaderSource(backend::ShaderModel shaderModel, uint8
+    assert_invariant(mMaterialTag == ChunkType::MaterialSpirv);
+
+    const EShLanguage shLang = stage == VERTEX ? EShLangVertex : EShLangFragment;
+
+    std::string nullTerminated(source, sourceLength);
+    source = nullTerminated.c_str();
+
+    TShader tShader(shLang);
+    tShader.setStrings(&source, 1);
+
+    MaterialBuilder::TargetApi targetApi = targetApiFromBackend(mBackend);
+    assert_invariant(targetApi == MaterialBuilder::TargetApi::VULKAN);
+
+    const int langVersion = GLSLTools::glslangVersionFromShaderModel(shaderModel);
+    const EShMessages msg = GLSLTools::glslangFlagsFromTargetApi(targetApi);
+    const bool ok = tShader.parse(&DefaultTBuiltInResource, langVersion, false, msg);
+    if (!ok) {
+        slog.e << "ShaderReplacer parse:\n" << tShader.getInfoLog() << io::endl;
+        return false;
+    }
+
+    TProgram program;
+    program.addShader(&tShader);
+    const bool linkOk = program.link(msg);
+    if (!linkOk) {
+        slog.e << "ShaderReplacer link:\n" << program.getInfoLog() << io::endl;
+        return false;
+    }
+
+    // Unfortunately we need to use std::vector to interface with glslang.
+    vector<unsigned int> spirv;
+
+    SpvOptions options;
+    options.generateDebugInfo = true;
+    GlslangToSpv(*tShader.getIntermediate(), spirv, &options);
+
+    source = (const char*) spirv.data();
+    sourceLength = spirv.size() * 4;
+
+    slog.i << "Success re-generating SPIR-V. (" << sourceLength << " bytes)" << io::endl;
+
+    // Clone all chunks except Dictionary* and Material*.
+    stringstream sstream(string((const char*) cc.getData(), cc.getSize()));
+    stringstream tstream;
+    BlobIndex shaderIndex;
+    {
+        uint64_t type;
+        uint32_t size;
+        vector<uint8_t> content;
+        while (sstream) {
+            sstream.read((char*) &type, sizeof(type));
+            sstream.read((char*) &size, sizeof(size));
+            content.resize(size);
+            sstream.read((char*) content.data(), size);
+            if (ChunkType(type) == mDictionaryTag) {
+                shaderIndex.addDataBlobs(content.data(), size);
+                continue;
+            }
+            if (ChunkType(type) == mMaterialTag) {
+                shaderIndex.addShaderRecords(content.data(), size);
+                continue;
+            }
+            tstream.write((char*) &type, sizeof(type));
+            tstream.write((char*) &size, sizeof(size));
+            tstream.write((char*) content.data(), size);
+        }
+    }
+
+    // Append the new chunks for Dictionary* and Material*.
+    if (!shaderIndex.isEmpty()) {
+        shaderIndex.replaceShader(shaderModel, variant, stage, source, sourceLength);
+        shaderIndex.writeBlobsChunk(mDictionaryTag, tstream);
+        shaderIndex.writeShadersChunk(mMaterialTag, tstream);
+    }
+
+    // Copy the new package from the stringstream into a ChunkContainer.
+    const size_t size = tstream.str().size();
+    uint8_t* data = new uint8_t[size];
```

Comment:
Does this leak memory?

Reply:
Yeah it gets freed by the DebugServer, which isn't obvious so I added some comments in a new commit.  Thanks for pointing this out.

Reply:
I see, thanks!